docs/ffi/ffi.md

```diff
@@ -0,0 +1,467 @@
+In this case, we actually define two new FFI calls:
+
+1. `rms_norm_fwd` returns two outputs: (a) the "primal" result, and (b) the "residuals" which are used in the backwards pass.
+2. `rms_norm_bwd` takes the residuals and the output co-tangents, and returns the input co-tangents.
+
+We won't get into the details of the RMS normalization backwards pass, but take a look at the [C++ source code](rms_norm.cc) to see how these functions are implemented on the back end.
+The main point to emphasize here is that the "residual" computed has a different shape than the primal output, therefore, in the {func}`~jax.extend.ffi.ffi_call` to `res_norm_fwd`, the output type has two elements with different shapes.
+
+This custom derivative rule can be wired in as follows:
+
+```{code-cell} ipython3
+xla_client.register_custom_call_target(
+  "rms_norm_fwd", rms_norm_lib.rms_norm_fwd(), platform="cpu", api_version=1
+)
+xla_client.register_custom_call_target(
+  "rms_norm_bwd", rms_norm_lib.rms_norm_bwd(), platform="cpu", api_version=1
+)
+
+
+def rms_norm_fwd(x, eps=1e-5):
+  y, res = jex.ffi.ffi_call(
+    "rms_norm_fwd",
+    (
+      jax.ShapeDtypeStruct(x.shape, x.dtype),
+      jax.ShapeDtypeStruct(x.shape[:-1], x.dtype),
+    ),
+    x,
+    eps=np.float32(eps),
+    vectorized=True,
+  )
+  return y, (res, x)
+
+
+def rms_norm_bwd(eps, res, ct):
+  del eps
+  res, x = res
+  assert res.shape == ct.shape[:-1]
+  assert x.shape == ct.shape
+  return (
+    jex.ffi.ffi_call(
+      "rms_norm_bwd",
+      jax.ShapeDtypeStruct(ct.shape, ct.dtype),
+      res,
+      x,
+      ct,
+      vectorized=True,
+    ),
+  )
+
+
+rms_norm = jax.custom_vjp(rms_norm, nondiff_argnums=(1,))
+rms_norm.defvjp(rms_norm_fwd, rms_norm_bwd)
+
+# Check that this gives the right answer when compared to the reference version
+ct_y = jnp.ones_like(x)
+np.testing.assert_allclose(
+  jax.vjp(rms_norm, x)[1](ct_y), jax.vjp(rms_norm_ref, x)[1](ct_y), rtol=1e-5
+)
+```
+
+At this point, we can use our new `rms_norm` function transparently for many JAX applications, and it will transform appropriately under the standard JAX function transformations like {func}`~jax.vmap` and {func}`~jax.grad`.
+One thing that this example doesn't support is forward-mode AD ({func}`jax.jvp`, for example) since {func}`~jax.custom_vjp` is restricted to reverse-mode.
+JAX doesn't currently expose a public API for simultaneously customizing both forward-mode and reverse-mode AD, but such an API is on the roadmap, so please [open an issue](https://github.com/google/jax/issues) describing you use case if you hit this limitation in practice.
+
+## FFI calls on a GPU
+
+So far, we have been interfacing only with foreign functions running on the CPU, but JAX's FFI also supports calls to GPU code.
+Since this documentation page is automatically generated on a machine without access to a GPU, we can't execute any GPU-specific examples here, but we will go over the key points.
```

Comment:
I thought that this would not be runnable under `pytest doctest`, as we do for other .md files, but now I am wondering what kind of automation do you have. If it is not automatic, it would be good to add some comments somewhere for how a JAX developer can refresh the documentation.

Reply:
`doctest` doesn't actually test most of the ipynb/md files in `docs` since AFAIK it only tests the specific syntax that includes `>>>`. But, like many of the other notebooks, this document is automatically executed on Read the Docs when generating the docs page and the custom calls get compiled there because RtDs comes with the all the necessary developer tools. There's nothing special about this tutorial compared to the others.

One nice benefit is that this means that this tutorial gives us a (non-blocking) CI test for the user version of the FFI API!

Reply:
Nice. Non-blocking has the disadvantage that somebody has to watch for breakages after submit :-(

Reply:
Yes! I am planning to add something like this more directly to CI, but it is somewhat involved so I haven't gotten to it yet. It's on the to do list!